src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+            return false;
+        }
+
+        private void ParseHeaderLineFast(ReadOnlySpan<byte> line,
+          ref KeyValueAccumulator accumulator)
+        {
+            ReadOnlySpan<byte> key;
+            ReadOnlySpan<byte> value;
+
+            int colon = line.IndexOf(ColonDelimiter);
+
+            if (colon == -1)
+            {
+                throw new InvalidDataException($"Invalid header line: {GetTrimmedString(line)}");
+            }
+
+            key = line.Slice(0, colon);
+            value = line.Slice(colon + ColonDelimiter.Length);
+
+            var decodedKey = GetTrimmedString(key);
+            var decodedValue = GetTrimmedString(value);
+            AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+        }
+
+        // For multi-segment parsing of a read only sequence
+        private bool TryParseHeadersToEndSlow(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            ref long headersLength)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> line, CrlfDelimiter))
+                {
+                    // Don't buffer indefinitely
+                    if (headersLength + sequenceReader.Consumed > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
+                    }
+
+                    return false;
+                }
+
+                if (sequenceReader.Consumed + headersLength > HeadersLengthLimit)
+                {
+                    throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                }
+
+                if (line.IsEmpty)
+                {
+                    buffer = buffer.Slice(sequenceReader.Position);
+                    headersLength += sequenceReader.Consumed;
+                    return true;
+                }
+
+                if (line.IsSingleSegment)
+                {
+                    ParseHeaderLineFast(line.FirstSpan, ref accumulator);
+                    continue;
+                }
+
+                var lineReader = new SequenceReader<byte>(line);
+                ReadOnlySequence<byte> value;
+
+                if (!lineReader.TryReadTo(out ReadOnlySequence<byte> key, ColonDelimiter))
+                {
+                    throw new InvalidDataException($"Invalid header line: {GetStringFromReadOnlySequence(line)}");
+                }
+                value = line.Slice(lineReader.Position);
+
+
+                var decodedKey = GetStringFromReadOnlySequence(key);
+                var decodedValue = GetStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+
+            buffer = default;
```

Comment:
Does this code path only execute when the sequence is empty?

Reply:
It executes when the buffer is empty, or if we read till the end of the buffer, but didn't find all the headers - and didn't exceed the Headers' length limit.